watcher.go

```diff
@@ -596,6 +601,26 @@ func (t *tracker) registerNotifiers(ns ...Notifier) error {
 	return nil
 }
 
+// deregisterNotifiers removes the notifiers from those tracked
+func (t *tracker) deregisterNotifiers(ns ...Notifier) {
```

Comment:
it looks like the associated views may not be handled yet, perhaps `tracker.markForSweep` and `tracker.sweep` may be useful for further cleanup?

Reply:
We talked about this a bit, how it could leak views. I asked Michael to leave it out for now because I wanted to look at it and it would be called by the resolver.Run normally anyways and that would take care of it.

Reply:
I wanted to review the markForSweep + sweep idea, thinking about races or issues with the locking. Upon review I don't think there are any issues with it. Each step is independent and idempotent (within the lock). It is 3 locks where it could be 1, but doing anything about this seems premature.

So my suggestion at this point would be to just have the deregisterNotifiers mirror the registerNotifiers and only delete it from the tracker.notifiers map. Then call tracker.markForSweep and tracker.sweep from the watcher.Deregister call. 

I'd prefer it be more self contained in the tracker, but the sweep needs to clear the cache which needs to be passed in from the watcher at this point.

Reply:
So just to clarify, remove the functionality I ported from sweep and instead in CTS it would look like what I sent you yesterday?

``` go
func (tf *Terraform) deleteTask(ctx context.Context) {
	tf.watcher.DeRegister(tf.template)

	// cleanup old template from watcher
	tf.watcher.MarkForSweep(tf.template)
	tf.watcher.Sweep(tf.template)
}
```

Reply:
I mean just have the Deregister() method on the watcher do those 3 steps. So CTS would just call `tf.Deregister(tf.template)` and that would the 3 methods on the internal tracker object.

Reply:
Added note... I don't want the mark/sweep stuff public. I want to keep that internal.

Reply:
Ok, so this:

```Go
func (w *Watcher) Deregister(ns ...Notifier) {
	w.tracker.deregisterNotifiers(ns...)

        for _, n := range ns{
                w.tracker.markForSweep(n)
                w.tracker.Sweep(n, w.cache)
        }
}
```

What do you mean by noting making the mark/sweep stuff public? They already have public methods?

Reply:
Sorry... you're getting a preview into my head. I was thinking about making them private.

I had them as public because I was thinking that you might want to resolve templates w/o the resolver and to have mark/sweep available in that case they'd need to be public. But I'm thinking this was a mistake exposing this lower level functionality and that the resolver should be the supported way to get that.